includes/schemas/class-prompts.php

```diff
@@ -0,0 +1,283 @@
+						'background_color'               => [
+							'name'     => 'background_color',
+							'type'     => 'string',
+							'format'   => 'hex-color',
+							'required' => false,
+							'default'  => '#FFFFFF',
+						],
+						'display_title'                  => [
+							'name'     => 'display_title',
+							'type'     => 'boolean',
+							'required' => false,
+							'default'  => false,
+						],
+						'hide_border'                    => [
+							'name'     => 'hide_border',
+							'type'     => 'boolean',
+							'required' => false,
+							'default'  => false,
+						],
+						'large_border'                   => [
+							'name'     => 'large_border',
+							'type'     => 'boolean',
+							'required' => false,
+							'default'  => false,
+						],
+						'frequency'                      => [
+							'name'     => 'frequency',
+							'type'     => 'string',
+							'required' => true,
+							'enum'     => [
+								'once',
+								'weekly',
+								'daily',
+								'always',
+								'custom',
+							],
+
+						],
+						'frequency_max'                  => [
+							'name'     => 'frequency_max',
+							'type'     => 'integer',
+							'required' => false,
+							'default'  => 0,
+						],
+						'frequency_start'                => [
+							'name'     => 'frequency_start',
+							'type'     => 'integer',
+							'required' => false,
+							'default'  => 0,
+						],
+						'frequency_between'              => [
+							'name'     => 'frequency_between',
+							'type'     => 'integer',
+							'required' => false,
+							'default'  => 0,
+						],
+						'frequency_reset'                => [
+							'name'     => 'frequency_reset',
+							'type'     => 'string',
+							'required' => false,
+							'default'  => 'month',
+							'enum'     => [
+								'month',
+								'week',
+								'day',
+							],
+						],
+						'overlay_color'                  => [
+							'name'     => 'overlay_color',
+							'type'     => 'string',
+							'required' => false,
+							'format'   => 'hex-color',
+							'default'  => '#000000',
+						],
+						'overlay_opacity'                => [
+							'name'     => 'overlay_opacity',
+							'type'     => 'integer',
+							'required' => false,
+							'default'  => 30,
+							'maximum'  => 100,
+						],
+						'overlay_size'                   => [
+							'name'     => 'overlay_size',
+							'type'     => 'string',
+							'required' => false,
+							'default'  => 'medium',
+							'enum'     => [
+								'x-small',
+								'small',
+								'medium',
+								'large',
+								'full-width',
+							],
+						],
+						'no_overlay_background'          => [
+							'name'     => 'no_overlay_background',
+							'type'     => 'boolean',
+							'required' => false,
+							'default'  => false,
+						],
+						'placement'                      => [
+							'name'     => 'placement',
+							'type'     => 'string',
+							'required' => true,
+							'enum'     => [
```

Comment:
Non-blocking: There are some other properties which interact with the various placement options. For example, if the prompt's placement is an overlay placement, then the `trigger_type` property cannot be `blocks_count`, which applies to inline prompts only. These conflicts should be handled by the [`Newspack_Popups_Model::generate_popup()`](https://github.com/Automattic/newspack-popups/blob/master/includes/class-newspack-popups-model.php#L1125) and [`generate_inline_popup()`](https://github.com/Automattic/newspack-popups/blob/master/includes/class-newspack-popups-model.php#L1062) methods (which will set reasonable default behaviors if they encounter unexpected property values), but wanted to state here that the schema makes it possible to create impossible configs, so we'll want to make sure we handle any possible conflicts upon prompt creation.

Reply:
Yes, I saw that.

The methods you link are called when the popup is being rendered, right? So they could not be used for validation...

In any case, I think we can handle that in a `sanitize` method in the Schema class. Before saving or exporting, we check all these internal dependencies and fix them... we could add these checks to the `is_valid` method as well...

Reply:
Yes, those methods will render the actual markup. I think it's okay to proceed for now as any "impossible" configs imported via JSON will be fixed by these methods when the prompts are rendered, but I agree that abstracting them to a `sanitize` or `is_valid` method is probably the better way to go, long-term.